uamqp/client.py

```diff
@@ -373,12 +373,25 @@ def _transfer_message(self, message_delivery, timeout=0):
     def _on_send_complete(self, message_delivery, message, reason, state):
         # TODO: check whether the callback would be called in case of message expiry or link going down
         # and if so handle the state in the callback
-        if SEND_DISPOSITION_ACCEPT in state:
+        if state and SEND_DISPOSITION_ACCEPT in state:
```

Comment:
In what case will state be None?

Reply:
In current implementation, it would happen in the following cases when we decide to discard all pending deliveries:

- we initiate detach
- service initiates the detach (a detach frame arrives)
- session state changed to DISCARDING (the connection got CLOSE_RCVD/END)

Reply:
hmm it feels like a bug that state should ever be None.... if this is the only place where we set it to none, maybe with can plug the gap with whatever the initial state is - or does that make the logic more difficult on line 381?

Reply:
according to the definition of spec, the delivery state (within a disposition frame) is used to inform the remote peer of local changes in the state of deliveries according to the spec.

and there're five types of state: `accepted/rejected/released/modified/received`.

according to the definition I think we'd better keep state as None to imply that it is a sender-side error in the case delivery not being sent at all, assigning a state as one of the aforementioned ones makes me feel we're changing the default behavior.

Reply:
I have added a TODO